File: acmeshop/acmeshop/src/main/java/org/acme/shop/ShopGraphQLResource.java
package org.acme.shop;

import io.smallrye.graphql.api.Subscription;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.operators.multi.processors.BroadcastProcessor;
import org.acme.shop.model.Customer;
import org.acme.shop.model.OrderItem;
import org.acme.shop.model.OrderItemInput;
import org.acme.shop.model.Product;
import org.acme.shop.model.ShopOrder;
import org.eclipse.microprofile.graphql.GraphQLApi;
import org.eclipse.microprofile.graphql.Mutation;
import org.eclipse.microprofile.graphql.Query;

import jakarta.annotation.PostConstruct;
import jakarta.transaction.Transactional;
import jakarta.validation.Valid;
import java.util.List;

@GraphQLApi
public class ShopGraphQLResource {

    @Query
    public List<Customer> getCustomers() {
        return Customer.listAll();
    }

    @Query
    public List<ShopOrder> getOrders() {
        return ShopOrder.listAll();
    }
    @Query
    public List<Product> getProducts() {
        return Product.listAll();
    }

    @Mutation
    @Transactional
    public Customer createCustomer(@Valid Customer customer) {
        customer.persist();
        return customer;
    }

    @Mutation
    @Transactional
    public ShopOrder createOrder(Long customerId, List<OrderItemInput> itemInput) {
        ShopOrder order = new ShopOrder();
        List<OrderItem> items = itemInput.stream()
            .map(input -> {
                OrderItem item = new OrderItem();
                item.setProduct(Product.findById(input.getProductId()));
                item.setQuantity(input.getQuantity());
                item.persistAndFlush();
                return item;
            }).toList();
        order.setItems(items);
        order.persistAndFlush();
        Customer.<Customer>findById(customerId).addOrder(order);
        newOrdersPublisher.onNext(order);
        return order;
    }

    private Multi<ShopOrder> newOrdersMulti;
    private BroadcastProcessor<ShopOrder> newOrdersPublisher;

    @PostConstruct
    public void initializeNewOrdersMulti() {
        newOrdersPublisher = BroadcastProcessor.create();
        newOrdersMulti = Multi.createFrom().publisher(newOrdersPublisher);
    }

    @Subscription
    public Multi<ShopOrder> newOrders() {
        return newOrdersMulti;
    }
}

File: acmeshop/acmeshop-kotlin/src/main/kotlin/shop/ShopGraphQLResource.kt
package shop

import io.smallrye.graphql.api.Subscription
import io.smallrye.mutiny.Multi
import io.smallrye.mutiny.operators.multi.processors.BroadcastProcessor
import jakarta.annotation.PostConstruct
import jakarta.transaction.Transactional
import jakarta.validation.Valid
import org.eclipse.microprofile.graphql.GraphQLApi
import org.eclipse.microprofile.graphql.Mutation
import org.eclipse.microprofile.graphql.Query
import shop.model.*
import java.util.function.Function

@GraphQLApi
class ShopGraphQLResource {
    @Query
    fun customers(): List<Customer> {
        return Customer.listAll()
    }

    @Query
    fun orders(): List<ShopOrder> {
        return ShopOrder.listAll()
    }


    @Query
    fun products(): List<Product> {
        return Product.listAll() as List<Product>
    }

    @Mutation
    @Transactional
    fun createCustomer(@Valid customer: Customer): Customer {
        customer.persist()
        return customer
    }

    @Mutation
    @Transactional
    fun createOrder(customerId: Long, itemInput: List<OrderItemInput>): ShopOrder {
        val order = ShopOrder()
        val items: List<OrderItem> = itemInput.stream()
                .map(Function<OrderItemInput, OrderItem> { input: OrderItemInput ->
                    val item = OrderItem()
                    item.product = Product.findById(input.productId!!)
                    item.quantity = input.quantity
                    item.persistAndFlush()
                    item
                }).toList()
        order.setItems(items.toMutableList())
        order.persistAndFlush()

        Customer.findById(customerId)!!.addOrder(order)
        newOrdersPublisher.onNext(order)
        return order
    }

    private val newOrdersPublisher: BroadcastProcessor<ShopOrder> = BroadcastProcessor.create()
    private val newOrdersMulti: Multi<ShopOrder> =  Multi.createFrom().publisher(newOrdersPublisher)

    @Subscription
    fun newOrders(): Multi<ShopOrder> {
        return newOrdersMulti
    }
}